worldedit-core/src/main/java/com/fastasyncworldedit/core/function/mask/AngleMask.java

```diff
@@ -126,38 +128,42 @@ protected boolean testSlope(Extent extent, int x, int y, int z) {
         }
     }
 
-    private boolean adjacentAir(Extent extent, MutableBlockVector3 mutable) {
-        int x = mutable.x();
-        int y = mutable.y();
-        int z = mutable.z();
-        if (!mask.test(extent, mutable.setComponents(x + 1, y, z))) {
+    // for optimal performance, this method should be called with base being a CharFilterBlock
+    private boolean adjacentAir(Extent extent, BlockVector3 base) {
+        int y = base.y();
+        // we expect the data for blocks above and below to be loaded already
+        // in which case caching/reading from cache has more overhead
+        if (y != maxY && !fastMask.test(base.getStateRelativeY(extent, 1))) {
             return true;
         }
-        if (!mask.test(extent, mutable.setComponents(x - 1, y, z))) {
+        if (y != minY && !fastMask.test(base.getStateRelativeY(extent, -1))) {
             return true;
         }
-        if (!mask.test(extent, mutable.setComponents(x, y, z + 1))) {
+        // other positions might be in different chunks, go a slower, cached path there
+        MutableBlockVector3 mutable = new MutableBlockVector3();
+        int x = base.x();
+        int z = base.z();
+        if (!mask.test(extent, mutable.setComponents(x + 1, y, z))) {
             return true;
         }
-        if (!mask.test(extent, mutable.setComponents(x, y, z - 1))) {
+        if (!mask.test(extent, mutable.setComponents(x - 1, y, z))) {
             return true;
         }
-        if (y != maxY && !mask.test(extent, mutable.setComponents(x, y + 1, z))) {
+        if (!mask.test(extent, mutable.setComponents(x, y, z + 1))) {
             return true;
         }
-        return y != minY && !mask.test(extent, mutable.setComponents(x, y - 1, z));
+        return !mask.test(extent, mutable.setComponents(x, y, z - 1));
     }
 
     @Override
     public boolean test(BlockVector3 vector) {
 
-        if (!mask.test(vector)) {
+        if (!fastMask.test(vector)) {
```

Comment:
I think we should cache this at least if on the edge of the chunk, as there's little point using the cache mask on x/z +/- 1 if there's unlikely to be a value already in the cache. We know we're going to be testing every block here

Reply:
Not sure if it's worth the complexity to have different code paths for those cases. I tried again to revert this change to use the cached mask and it makes the whole mask slower even in the overlay scenario.

I also replaced the usages of the `CachedMask` in `adjacentAir`, bringing time spent in `adjacentAir` down from ~25s (when using the cached mask everywhere) to ~10s in my test... so I guess we really need to look into CachedMask after your PR.

Reply:
Okay that's definitely concerning then. I suppose the caching makes it slower if the chunks are already available, and we will already have those GET chunks loaded in FAWE

Reply:
Yes that code path doesn't seem to run into chunk loading, so any access is more or less direct (going through STQE)